hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -113,6 +113,45 @@ class LowerTableIR(val typesToLower: DArrayLowering.Type) extends AnyVal {
             throw new LowererUnsupportedOperation(s"can't lower a TableRead with reader $r.")
         }
 
+      case TableParallelize(rowsAndGlobal, nPartitions) =>
+        val nPartitionsAdj = nPartitions.getOrElse(16)
+        val loweredRowsAndGlobal = lowerIR(rowsAndGlobal)
+
+        val contextType = TStruct(
+          "elements" -> TArray(GetField(loweredRowsAndGlobal, "rows").typ.asInstanceOf[TArray].elementType)
+        )
+
+        val context = MakeStream((0 until nPartitionsAdj).map { partIdx =>
+          val numRows = ArrayLen(GetField(loweredRowsAndGlobal, "rows"))
+          val numNonEmptyPartitions = If(numRows < nPartitionsAdj, numRows, nPartitionsAdj)
+          val q = numRows floorDiv numNonEmptyPartitions
+          val remainder = numRows - q * numNonEmptyPartitions
+          val length = If(numNonEmptyPartitions >= partIdx,
+            If(remainder > 0,
+              If(remainder > partIdx, q + 1, q),
+              q),
+            0
+          )
+          val start = If(numNonEmptyPartitions >= partIdx,
+            If(remainder > 0,
+              If(remainder < partIdx, q * partIdx + remainder, (q + 1) * partIdx),
+              q * partIdx
+            ),
+            0
+          )
+          val elements = ToArray(StreamTake(StreamDrop(ToStream(GetField(loweredRowsAndGlobal, "rows")), start), length))
```

Comment:
I think this evaluates loweredRowsAndGlobals twice.  I think you need to put the whole thing in the table stage globals, and then pull out the rows from there.

Reply:
Also, StreamTake/StreamDrop are O(N), which makes this quadratic.  You should do something like `StreamMap(StreamRange(...), "i", ArrayRef(Ref("rows"), Ref("i")))`.

Reply:
> I think this evaluates loweredRowsAndGlobals twice. I think you need to put the whole thing in the table stage globals, and then pull out the rows from there.


Then the rows are sent to every partition, right? John and I discussed this and decided duplication might be a better choice, but happy to be corrected.

Reply:
Shouldn't be.  Only the context goes to the partition, and the context should only have the elements destined for that partition.  Either way, you can't be evaluating this twice.

It might be useful to print the lowered code for an example.  We should be doing more of this.

Reply:
I'll move it to the globals, that's fine. But don't things that go to globals get sent to every executor? So I'll end up copying the whole array to each executor? That's probably fine, you can't use parallelize on something really big anyway, I just wasn't sure whether we were better off evaluating it twice or sending extra data.

Reply:
Ah, I think this is a deficiency in TableStage.  I think it needs to carry three things:
 - things that should be bound by a let for downstream consumption,
 - things that should be broadcast,
 - and the globals within the broadcast values.

The non-broadcast bindings should be a List[(String, IR)] where earlier elements can refer to later ones in the list.

Then Parallelize evaulates the value in the bindings, refers to the globals in the broadcast value, and uses the non-broadcast arrays when building the context.

The non-broadcast bindings should be visible to all globals and the context.  So a TableStage renders down to:

```
(Let* non-broadcast-bindings
  (CDA broadastVals contexts ...))
```

I'm not 100% happy  with broadcastVals being a single IR.  I think this will lead to a lot of nested MakeStructs that aren't particularly efficient.  But let's address this later.

Reply:
Sure, this makes sense. 

So in the interim, the thing to do is to just to put everything in the globals that are going to be broadcast in order to avoid evaluating the IR twice? I've got a branch in progress where I've addressed your other comments.

Reply:
#8599